File: leetcode/binary_tree_level_order_traversal/BinaryTreeLevelOrderTraversal.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int> > levelOrder(TreeNode *root) {
        // Note: The Solution object is instantiated only once and is reused by each test case.
        vector<vector<int> > result; 
        if (!root) {
            return result; 
        }
        
        TreeNode * prev = NULL; 
        TreeNode * curr = NULL; 
        queue<TreeNode *> q; 
        q.push(root);
        
        while (q.size() != 0) {
            prev = curr; 
            curr = q.front();
            q.pop(); 
            
            if (prev == NULL) {
                q.push(NULL);
                vector<int> temp; 
                result.push_back(temp); 
            }
            
            if (curr != NULL) {
                result[result.size() - 1].push_back(curr->val);
                if (curr->left != NULL) {
                    q.push(curr->left); 
                }
                
                if (curr->right != NULL) {
                    q.push(curr->right); 
                }
            }
        }
        
        return result; 
    }
};

File: leetcode6/binary_tree_level_order_traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 * 
 * Binary tree level order traversal. Traverse the binary tree level by level and 
 * return the values of each level in a list.
 * Two levels alternating traversal approach. Each time store the value of current 
 * level to the result. 
 */
public class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> result = new ArrayList<List<Integer>>(); 
        List<List<TreeNode>> levels = new ArrayList<List<TreeNode>>(); 
        levels.add(new ArrayList<TreeNode>()); 
        levels.add(new ArrayList<TreeNode>()); 
        int lev0 = 0; 
        int lev1 = 1; 
        
        if (root != null) {
            levels.get(lev0).add(root); 
        }
        
        while (levels.get(lev0).size() != 0) {
            List<TreeNode> curr = levels.get(lev0); 
            List<TreeNode> next = levels.get(lev1);
            List<Integer> temp = new ArrayList<>(); 
            
            for (TreeNode item : curr) {
                temp.add(item.val); 
                
                if (item.left != null) {
                    next.add(item.left); 
                }
                
                if (item.right != null) {
                    next.add(item.right); 
                }
            }
            
            result.add(temp); 
            curr.clear(); 
            lev0 = lev0 ^ lev1; 
            lev1 = lev0 ^ lev1; 
            lev0 = lev0 ^ lev1; 
        }
        
        return result; 
    }
}
